docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+
+   ```
+   tedge/<child-id>/device/commands/req/software_update
+   tedge/<child-id>/device/commands/res/software_update
+   ```
+
+* Firmware update operation
+
+   ```
+   tedge/main/device/commands/req/firmware_update
+   tedge/main/device/commands/res/firmware_update
+   ```
+
+* Device restart operation
+
+   ```
+   tedge/main/device/commands/req/device_restart
+   tedge/main/device/commands/res/device_restart
+   ```
+
+* Configuration snapshot operation
+
+   ```
+   tedge/main/service/<service-id>/commands/req/config_snapshot
+   tedge/main/service/<service-id>/commands/res/config_snapshot
+   ```
+
+* Configuration update operation
+
+   ```
+   tedge/<child-id>/service/<service-id>/commands/req/config_update
+   tedge/<child-id>/service/<service-id>/commands/res/config_update
+   ```
+
+Although all the above examples maintain consistent structure by ending with the `<operation-type>`,
+further additions are possible in future if desired for a given operation type.
+For e.g: `tedge/main/commands/req/config_update/<config-type>` to address a specific `config-type`
+
+### Registration service for child devices
+
+Immediate and nested child devices can be registered with thin-edge using its registration service,
+by sending the following MQTT message to the topic: `tedge/main/register/req/child`:
+
+```json
+{ 
+   "parent": "<parent-device-id>",
+   "id-prefix": "<desired-child-id-prefix>",
+   "capabilities": {
+      "<capability-1>": {},  //capability-1 specific metadata
+      "<capability-2>": {},  //capability-2 specific metadata
+      ...
+   }
+}
+```
+
+The `parent-device-id` is the device-id of the direct parent that the child device is connected to.
+The payload can have other fields describing the capabilities of that device as well (config management, software management etc).
+
+```admonish warning
+The exact topic keys and payload format for this init contract can be discussed and refined separately.
+The focus here is just on the MQTT topic structure.
+```
+
+Thin-edge needs to maintain the lineage (hierarchy of parent devices) of all descendent child devices in its internal state,
+so that it can be looked up while receiving any data from them.
+Even though the child device only declares its immediate parent in the registration message,
+the entire lineage can be traced back with a recursive lookup on that `parent-device-id`.
+
+The registration status, whether the device registration succeeded or not, is sent back on `tedge/main/register/res/child`,
+with the internal device id used by thin-edge to uniquely identify that device as follows:
+
+```json
+{
+   "id": "<generated-id>",
+   "status": "successful" 
+}
+```
+
+The `<generated-id>` will use the `<desired-child-id-prefix>` sent in the registration request.
+
+A failure is indicated with a failed status: `{ "status": "failed" }`.
+
+Once the registration is complete, these nested child devices can use the `tedge/<generated-id>` topic prefix
```

Comment:
*can* or *must* ?

Reply:
I would still like to keep the registration step optional for device fleets that can guarantee uniqueness. So, in such cases, even if they are going through the optional registration step (not for unique id generation, but just to get other metadata registered, like supported operations), they should be allowed to use their own `id` in the topics.

To support that case, I'll add support for a `desired-id` field as well in the registration message payload, in addition to the `id-prefix` field.

Reply:
Though in this context `should` would be better, as if the child device has already gone through the registration process, then it `should` use the given generated-id if wants any data to be linked to the entity created as part of the registration process.